File: pangrams/pangrams.cpp
#include <algorithm>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <locale>
#include <string>
#include <ctime>



const std::string ALPHABET = "abcdefghijklmnopqrstuvwxyz";


void ToLower(std::string &text) {
    std::locale locale;
    std::transform(text.begin(), text.end(), text.begin(), [&locale](char sym) { return std::tolower(sym, locale); });
}


void RunAndMeasure(const std::function<bool(const std::string &)> &func, int count, std::string text) {
    ToLower(text);
    std::cout << (func(text) ? "true" : "false") << ' ';
    std::clock_t start = std::clock();
    for (int i = 0; i < count; ++i)
        func(text);
    std::clock_t finish = std::clock();
    std::cout << (finish - start) << std::endl;
}


bool CheckForEveryDigit(const std::string &text) {
    for (char sym : ALPHABET) {
        std::string lowText(text);
        ToLower(lowText);
        if (lowText.find(sym) == std::string::npos)
            return false;
    }
    return true;
}


bool CheckForEveryDigitWithoutLower(const std::string &text) {
    for (char sym : ALPHABET) {
        if (text.find(sym) == std::string::npos)
            return false;
    }
    return true;
}


bool CheckViaFlags(const std::string &text) {
    int flags = 0;
    for (char sym : text) {
        unsigned char code = sym - 'a';
        if (code < 26)
            flags |= (1 << code);
    }
    return (flags == (1 << 26) - 1);
}


void RunAll(int count, const std::string &text) {
    std::cout << "Text size: " << text.size() << std::endl;
    //RunAndMeasure(CheckForEveryDigit, count, text);
    RunAndMeasure(CheckForEveryDigitWithoutLower, count, text);
    RunAndMeasure(CheckViaFlags, count, text);
    std::cout << std::endl;
}


int main() {
    RunAll(10000, "Jackdaws love my big sphinx of quartz");
    //RunAll(10000, "Jackdaws love my sphinx of quartz");
    std::string text;
    for (int i = 0; i < 1000000; ++i) {
        text += ' ';
        if (rand() % 2 == 0)
            text += '.';
        else
            text += ',';
    }
    for (int i = 0; i < ALPHABET.size(); ++i) {
        for (int j = 10; j > 0; --j) {
            text += ALPHABET[i];
        }
    }
    RunAll(1, text);
    return 0;
}

File: pangrams/pangrams.py
import time


alphabet = 'abcdefghijklmnopqrstuvwxyz'


def run_and_measure(func, count, text):
    print(func(text))
    start = time.time()
    for _ in range(count):
        func(text)
    finish = time.time()
    print(finish - start)
    print('')


def check_for_every_digit(text):
    for sym in alphabet:
        if sym not in text.lower():
            print(sym, text)
            exit(0)
            return False
    return True


def check_for_every_digit_without_lower(text):
    text = text.lower()
    for sym in alphabet:
        if sym not in text:
            return False
    return True


def check_via_flags(text):
    text = text.lower()
    flags = [False for _ in range(26)]
    for sym in text:
        code = ord(sym) - ord('a')
        if 0 <= code < 26:
            flags[code] = True
    return all(flags)


def check_via_set(text):
    used = set()
    for sym in text.lower():
        if 'a' <= sym <= 'z':
            used.add(sym)
    return len(used) == 26


def run_all(count, text):
    print(len(text))
    run_and_measure(check_for_every_digit, count, text)
    run_and_measure(check_for_every_digit_without_lower, count, text)
    run_and_measure(check_via_flags, count, text)
    run_and_measure(check_via_set, count, text)


def main():
    run_all(10000, 'Jackdaws love my big sphinx of quartz')
    text = ''
    for i in range(len(alphabet)):
        for _ in range(1000000 if i < 2 else 10):
            text += alphabet[i]
    run_all(1, text)


if __name__ == '__main__':
    main()

